README
ExternalArticles
----------------

A MediaWiki extension for loading page text from external sources

https://www.mediawiki.org/wiki/Extension:ExternalArticles


CHANGES
== 2017-02-21 version 0.4.0 ==
* Add system for importing text files.

== 2017-02-21 version 0.3.0 ==
* Switch to use the extension registration system (and therefore increase MediaWiki version requirement to 1.24).

== 2013-05-31 version 0.1.4 ==
* Updated to use the MWHttpRequest class (increasing the MediaWiki version
  requirement to 1.17 when this class was introduced).
* Code formatting.
* Code moved to Git.
* Extension adopted by User:Samwilson (with Nate's permission).
* In-code license updated to GPL version 3 (to be consistent with the licence
  named elsewhere).

== 2010-03-05 version 0.1.3 ==
* Removed dependence on PHP config "allow_url_fopen" by changing to use cURL.
  Thanks to [[User:Alvinos|Alvinos]].

== Version 0.1.2 ==
* Rephrased errors, added todo's and other minor changes.

== Version 0.1.1 ==
* Modified the $wgExtensionCredits:
  Bug: hmm... defined the 'author' element twice when I meant 'url' the second
  time.
  Bug: The url specified was incorrect.
  Added a contact email address specific to this extension.

== Version 0.1 ==
* Preloads text into an edit box on article edit.
* Configure external wiki to draw text from (defaults to Wikipedia).
* Configure perl-compatible regular expression to include or exclude certain
  pages (defaults to Templates only).


src/Hooks.php
<?php

namespace MediaWiki\Extensions\ExternalArticles;

use MWHttpRequest;
use Title;

class Hooks {

	/**
	 * Preload text from a remote wiki into the edit form. Called when edit page for
	 * a new article is shown.
	 *
	 * @global OutputPage $wgOut OutputPage object for HTTP response
	 * @global array $eagRules ExternalArticles configuration array
	 * @param string &$text Text with which to prefill the edit form
	 * @param Title &$title Title of the new page
	 * @return bool
	 */
	function onEditFormPreloadText( &$text, Title &$title ) {
		global $wgOut, $wgEagRules;

		// @todo: change this so each setting is set to it's default if it is not defined.
		// Currently, if anything is overridden, all must be defined.
		if ( !isset( $wgEagRules ) || is_null( $wgEagRules ) ) {
			$wgEagRules = [];
			$wgEagRules['onpreload'] = true;
			$wgEagRules['url'] = 'https://en.wikipedia.org/w/index.php?title=';

			// @todo: remove assumption of English.
			$wgEagRules['rule'] = '/^Template:.*$/';
		} else {
			// @todo: validate $wgEagRules URL's, etc...
		}

		$pagename = $title->getPrefixedURL();
		$url = $wgEagRules['url'] . $pagename . '&action=raw';
		$ismatch = preg_match( $wgEagRules['rule'], $pagename ) > 0;

		if ( defined( 'EXTERNALARTICLES_DEBUG' ) ) {
			if ( $ismatch ) {
				$wgOut->addWikiText( "URL: $url<br />" );
			} else {
				$wgOut->addWikiText( "Page title does not match rule.<br />" );
			}
		}

		if ( $wgEagRules['onpreload'] && $ismatch && empty( $text ) ) {
			$options = [
				'followRedirects' => true,
			];
			$httpRequest = MWHttpRequest::factory( $url, $options );
			$status = $httpRequest->execute();
			if ( !$status->isOK() ) {
				if ( defined( 'EXTERNALARTICLES_DEBUG' ) ) {
					$wgOut->addWikiText( "Failed to fetch external page: " . $status->getWikiText() );
				}

				return false;
			}
			$wgOut->wrapWikiMsg( '<div class="success">$1</div>',
				[ 'externalarticles-article-loaded', $url ] );
			$text = $httpRequest->getContent();

			return true;
		}

		return true;
	}
}


src/TextFileImporter.php
<?php

namespace MediaWiki\Extensions\ExternalArticles;

use DirectoryIterator;
use Exception;
use MediaWiki\MediaWikiServices;
use RecentChange;
use Revision;
use Title;
use User;
use WikiRevision;

class TextFileImporter {

	/** @var resource */
	protected $inotify;

	/** @var string[] */
	protected $watches;

	/**
	 * Import text files from a directory.
	 * @param string $dir The directory to import from.
	 * @param bool $watch Whether to continue to watch the files for changes.
	 * @throws Exception If the directory does not exist.
	 */
	public function __construct( $dir, $watch ) {
		if ( !is_dir( $dir ) ) {
			throw new Exception( "'$dir' is not a directory" );
		}
		$this->dir = realpath( $dir );
		$this->watch = (bool)$watch && function_exists( 'inotify_init' );
		if ( $this->watch ) {
			$this->inotify = inotify_init();
		}
	}

	/**
	 *
	 */
	public function import() {
		$topLevel = new DirectoryIterator( $this->dir );
		foreach ( $topLevel as $file ) {
			if ( $file->isDot() ) {
				continue;
			}
			if ( $file->isDir() ) {
				// Use the directory names as namespaces.
				$secondLevel = new DirectoryIterator( $this->dir . '/' . $file );
				foreach ( $secondLevel as $subfile ) {
					if ( $subfile->isDot() ) {
						continue;
					}
					$this->importFile( $subfile->getPathname() );
				}
			} else {
				$this->importFile( $file->getPathname() );
			}
		}

		if ( $this->watch ) {
			while ( true ) {
				$events = inotify_read( $this->inotify );
				foreach ( $events as $event ) {
					$file = $this->watches[ $event['wd'] ];
					$this->importFile( $file );
				}
			}
		}
	}

	/**
	 * Import a single file.
	 * @param string $file Full filesystem path to the file to import.
	 * @return bool
	 */
	protected function importFile( $file ) {
		// Construct the page name from the last components of the file path.
		$pagePath = substr( $file, strlen( $this->dir ) + 1 );
		$pageName = str_replace( '/', ':', $pagePath );

		// Have to check for # manually, since it gets interpreted as a fragment
		$title = Title::newFromText( $pageName );
		if ( !$title || $title->hasFragment() ) {
			echo "Invalid title: $pageName\n";
			return false;
		}

		if ( $this->watch ) {
			// Watch this file.
			$watchId = inotify_add_watch( $this->inotify, $file, IN_MODIFY );
			$this->watches[ $watchId ] = $file;
		}

		$exists = $title->exists();
		$oldRevID = $title->getLatestRevID();
		$oldRev = $oldRevID ? Revision::newFromId( $oldRevID ) : null;
		$actualTitle = $title->getPrefixedDBkey();

		$text = file_get_contents( $file );

		$rev = new WikiRevision( MediaWikiServices::getInstance()->getMainConfig() );
		$rev->setText( rtrim( $text ) );
		$rev->setTitle( $title );
		$user = User::newSystemUser( 'Maintenance script', [ 'steal' => true ] );
		$rev->setUserObj( $user );
		$rev->setComment( 'Imported by ExternalArticles extension' );
		$rev->setTimestamp( wfTimestampNow() );

		if ( $exists && $rev->getContent()->equals( $oldRev->getContent() ) ) {
			echo "$actualTitle does not need to be updated\n";
			return false;
		}

		$status = $rev->importOldRevision();
		$newId = $title->getLatestRevID();

		if ( $status ) {
			$action = $exists ? 'updated' : 'created';
			echo "Successfully $action $actualTitle\n";
		} else {
			$action = $exists ? 'update' : 'create';
			echo "Failed to $action $actualTitle\n";
			return false;
		}

		// Create the RecentChanges entry if necessary
		if ( $exists ) {
			if ( is_object( $oldRev ) ) {
				$oldContent = $oldRev->getContent();
				RecentChange::notifyEdit(
					$rev->getTimestamp(),
					$title,
					$rev->getMinor(),
					$user,
					$rev->getComment(),
					$oldRevID,
					$oldRev->getTimestamp(),
					false,
					'',
					$oldContent ? $oldContent->getSize() : 0,
					$rev->getContent()->getSize(),
					$newId,
					1
				);
			}
		} else {
			RecentChange::notifyNew(
				$rev->getTimestamp(),
				$title,
				$rev->getMinor(),
				$user,
				$rev->getTimestamp(),
				false,
				'',
				$rev->getContent()->getSize(),
				$newId,
				1
			);
		}
	}
}


ExternalArticles.i18n.php
<?php
/**
 * This is a backwards-compatibility shim, generated by:
 * https://git.wikimedia.org/blob/mediawiki%2Fcore.git/HEAD/maintenance%2FgenerateJsonI18n.php
 *
 * Beginning with MediaWiki 1.23, translation strings are stored in json files,
 * and the EXTENSION.i18n.php file only exists to provide compatibility with
 * older releases of MediaWiki. For more information about this migration, see:
 * https://www.mediawiki.org/wiki/Requests_for_comment/Localisation_format
 *
 * This shim maintains compatibility back to MediaWiki 1.17.
 */
$messages = [];
if ( !function_exists( 'wfJsonI18nShim7ed12447d100dede' ) ) {
	function wfJsonI18nShim7ed12447d100dede( $cache, $code, &$cachedData ) {
		$codeSequence = array_merge( [ $code ], $cachedData['fallbackSequence'] );
		foreach ( $codeSequence as $csCode ) {
			$fileName = __DIR__ . "/i18n/$csCode.json";
			if ( is_readable( $fileName ) ) {
				$data = FormatJson::decode( file_get_contents( $fileName ), true );
				foreach ( array_keys( $data ) as $key ) {
					if ( $key === '' || $key[0] === '@' ) {
						unset( $data[$key] );
					}
				}
				$cachedData['messages'] = array_merge( $data, $cachedData['messages'] );
			}

			$cachedData['deps'][] = new FileDependency( $fileName );
		}
		return true;
	}

	$GLOBALS['wgHooks']['LocalisationCacheRecache'][] = 'wfJsonI18nShim7ed12447d100dede';
}


ExternalArticles.php
<?php
/**
 * ExternalArticles.php
 *
 * The ExternalArticles extension fetches pages from a remote wiki.
 *
 * Copyright (C) 2009-2013 the authors listed below.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
 * http://www.gnu.org/copyleft/gpl.html
 *
 * @ingroup Extensions
 * @author Nathan Perry <externalarticles@nateperry.org>
 * @author Alvinos https://www.mediawiki.org/wiki/User:Alvinos
 * @author Sam Wilson <sam@samwilson.id.au>
 * @link http://www.nateperry.org/wiki/External_Articles
 * @license http://www.gnu.org/copyleft/gpl.html GNU General Public License
 * @file
 */
/**
 * Protect against register_globals vulnerabilities.
 * This line must be present before any global variable is referenced.
 */
if ( !defined( 'MEDIAWIKI' ) ) {
	echo <<<EOT
This file is not a valid entry point.
To install this extension, put the following line in LocalSettings.php:
require_once( "\$IP/extensions/ExternalArticles/ExternalArticles.php" );
EOT;
	exit( 1 );
}

/**
 * Extension setup.
 */
$wgExtensionCredits['other'][] = array(
	'path' => __FILE__,
	'name' => 'External Articles',
	'descriptionmsg' => 'externalarticles-desc',
	'version' => '0.2.0', // version date 2014-03-31
	'author' => array( 'Nathan Perry', 'Alvinos', 'Sam Wilson' ),
	'url' => 'https://www.mediawiki.org/wiki/Extension:ExternalArticles'
);
$wgExtensionMessagesFiles['ExternalArticles'] = dirname( __FILE__ ) . '/ExternalArticles.i18n.php';
$wgHooks['EditFormPreloadText'][] = 'ExternalArticles_EditFormPreloadText';

$wgMessagesDirs['ExternalArticles'] = __DIR__ . '/i18n';
$wgExtensionMessagesFiles['ExternalArticles'] = __DIR__ . '/ExternalArticles.i18n.php';

// @todo: change this so each setting is set to it's default if it is not defined.
//       Currently, if anything is overridden, all must be defined.
if ( !isset( $eagRules ) || is_null( $eagRules ) ) {
	$eagRules = array();
	$eagRules['onpreload'] = true;
	$eagRules['url'] = 'https://en.wikipedia.org/w/index.php?title=';

	// @todo: remove assumption of English.
	$eagRules['rule'] = '/^Template:.*$/'; // http://us3.php.net/manual/en/function.preg-match.php
} else {
	// @todo: validate $eagRules URL's, etc...
}


/**
 * Preload text from a remote wiki into the edit form. Called when edit page for
 * a new article is shown.
 *
 * @global OutputPage $wgOut OutputPage object for HTTP response
 * @global array $eagRules ExternalArticles configuration array
 * @param string $text Text with which to prefill the edit form
 * @param Title $title Title of the new page
 * @return boolean
 */
function ExternalArticles_EditFormPreloadText( &$text, &$title ) {
	global $wgOut, $eagRules;

	$pagename = $title->getPrefixedURL();
	$url = $eagRules['url'] . $pagename . '&action=raw';
	$ismatch = preg_match( $eagRules['rule'], $pagename ) > 0;

	if ( defined( 'EXTERNALARTICLES_DEBUG' ) ) {
		if ( $ismatch ) {
			$wgOut->addWikiText( "URL: $url<br />" );
		} else {
			$wgOut->addWikiText( "Page title does not match rule.<br />" );
		}
	}

	if ( $eagRules['onpreload'] && $ismatch && empty( $text ) ) {
		$options = array(
			'followRedirects' => true,
		);
		$httpRequest = MWHttpRequest::factory( $url, $options );
		$status = $httpRequest->execute();
		if ( !$status->isOK() ) {
			if ( defined( 'EXTERNALARTICLES_DEBUG' ) ) {
				$wgOut->addWikiText( "Failed to fetch external page: " . $status->getWikiText() );
			}
			return false;
		}
		$wgOut->wrapWikiMsg('<div class="success">$1</div>', array('externalarticles-article-loaded', $url));
		$text = $httpRequest->getContent();

		return true;
	}
	return true;
}

